PagingWithNetworkSample/app/src/main/java/com/android/example/paging/pagingwithnetwork/reddit/ui/RedditActivity.kt

```diff
@@ -78,34 +80,38 @@ class RedditActivity : AppCompatActivity() {
         initSearch()
     }
 
+    override fun onDestroy() {
+        super.onDestroy()
+        refreshListener?.let { (list.adapter as? PostsAdapter)?.removeLoadStateListener(it) }
+        refreshListener = null
+    }
+
     private fun initAdapter() {
         val glide = GlideApp.with(this)
-        val adapter = PostsAdapter(glide) {
-            model.retry()
+        val adapter = PostsAdapter(glide)
+        list.adapter = adapter.withLoadStateHeaderAndFooter(
+                header = PostsLoadStateAdapter(adapter),
+                footer = PostsLoadStateAdapter(adapter)
+        )
+
+        refreshListener = { loadType: LoadType, loadState: LoadState ->
+            if (loadType == LoadType.REFRESH) {
+                swipe_refresh.isRefreshing = loadState == LoadState.Loading
+            }
+        }.also { refreshListener ->
+            adapter.addLoadStateListener(refreshListener)
         }
-        list.adapter = adapter
-        model.posts.observe(this, Observer<PagedList<RedditPost>> {
-            adapter.submitList(it) {
-                // Workaround for an issue where RecyclerView incorrectly uses the loading / spinner
-                // item added to the end of the list as an anchor during initial load.
-                val layoutManager = (list.layoutManager as LinearLayoutManager)
-                val position = layoutManager.findFirstCompletelyVisibleItemPosition()
-                if (position != RecyclerView.NO_POSITION) {
-                    list.scrollToPosition(position)
-                }
+
+        lifecycleScope.launch {
+            model.posts.collect {
+                adapter.presentData(it)
             }
-        })
-        model.networkState.observe(this, Observer {
-            adapter.setNetworkState(it)
-        })
+        }
     }
 
     private fun initSwipeToRefresh() {
-        model.refreshState.observe(this, Observer {
-            swipe_refresh.isRefreshing = it == NetworkState.LOADING
-        })
         swipe_refresh.setOnRefreshListener {
-            model.refresh()
+            (list.adapter as? PostsAdapter)?.refresh()
```

Comment:
store the adapter so you can call refresh directly without casting, since you're using it in multiple places

Reply:
Done